packages/react-reconciler/src/__tests__/ReactHooks-test.internal.js

```diff
@@ -0,0 +1,1960 @@
+          ReactNoop.yield(`Did commit [${props.count}]`);
+        });
+        return <Text text={'Count: ' + props.count} />;
+      }
+      ReactNoop.render(<Counter count={0} />);
+      expect(ReactNoop.flush()).toEqual(['Count: 0']);
+      expect(ReactNoop.getChildren()).toEqual([span('Count: 0')]);
+      flushPassiveEffects();
+      expect(ReactNoop.clearYields()).toEqual(['Did commit [0]']);
+
+      ReactNoop.render(<Counter count={1} />);
+      expect(ReactNoop.flush()).toEqual(['Count: 1']);
+      expect(ReactNoop.getChildren()).toEqual([span('Count: 1')]);
+      // Effects are deferred until after the commit
+      flushPassiveEffects();
+      expect(ReactNoop.clearYields()).toEqual(['Did commit [1]']);
+    });
+
+    it('flushes passive effects even with sibling deletions', () => {
+      function LayoutEffect(props) {
+        useLayoutEffect(() => {
+          ReactNoop.yield(`Layout effect`);
+        });
+        return <Text text="Layout" />;
+      }
+      function PassiveEffect(props) {
+        useEffect(() => {
+          ReactNoop.yield(`Passive effect`);
+        }, []);
+        return <Text text="Passive" />;
+      }
+      let passive = <PassiveEffect key="p" />;
+      ReactNoop.render([<LayoutEffect key="l" />, passive]);
+      expect(ReactNoop.flush()).toEqual(['Layout', 'Passive', 'Layout effect']);
+      expect(ReactNoop.getChildren()).toEqual([
+        span('Layout'),
+        span('Passive'),
+      ]);
+
+      // Destroying the first child shouldn't prevent the passive effect from
+      // being executed
+      ReactNoop.render([passive]);
+      expect(ReactNoop.flush()).toEqual(['Passive effect']);
+      expect(ReactNoop.getChildren()).toEqual([span('Passive')]);
+
+      // (No effects are left to flush.)
+      flushPassiveEffects();
+      expect(ReactNoop.clearYields()).toEqual(null);
+    });
+
+    it('flushes passive effects even if siblings schedule an update', () => {
+      function PassiveEffect(props) {
+        useEffect(() => {
+          ReactNoop.yield('Passive effect');
+        });
+        return <Text text="Passive" />;
+      }
+      function LayoutEffect(props) {
+        let [count, setCount] = useState(0);
+        useLayoutEffect(() => {
+          // Scheduling work shouldn't interfere with the queued passive effect
+          if (count === 0) {
+            setCount(1);
+          }
+          ReactNoop.yield('Layout effect ' + count);
+        });
+        return <Text text="Layout" />;
+      }
+      ReactNoop.render([<PassiveEffect key="p" />, <LayoutEffect key="l" />]);
+      expect(ReactNoop.flush()).toEqual([
+        'Passive',
+        'Layout',
+        'Layout effect 0',
+        'Passive effect',
+        'Layout',
+        'Layout effect 1',
+      ]);
+      expect(ReactNoop.getChildren()).toEqual([
+        span('Passive'),
+        span('Layout'),
+      ]);
+    });
+
+    it('flushes passive effects even if siblings schedule a new root', () => {
```

Comment:
I wonder how you come up with all of those edge cases. 😉

Reply:
Ran into bugs in product code.